src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java

```diff
@@ -0,0 +1,473 @@
+              final Set<JsonNode> irresolvableReferences =
+                  getIrresolvableReferences(product, idToKey);
+              return irresolvableReferences.isEmpty();
+            })
+        .collect(Collectors.toList());
+  }
+
+  private Set<JsonNode> getIrresolvableReferences(
+      @Nonnull final Product product, @Nonnull final Map<String, String> idToKey) {
+
+    return getAllReferences(product).stream()
+        .filter(reference -> !idToKey.containsKey(getId(reference)))
+        .collect(toSet());
+  }
+
+  private static void replaceReferences(
+      @Nonnull final List<JsonNode> references, @Nonnull final Map<String, String> idToKey) {
+
+    references.forEach(
+        reference -> {
+          final String id = reference.get(REFERENCE_ID_FIELD).asText();
+          final String key = idToKey.get(id);
+          ((ObjectNode) reference).put(REFERENCE_ID_FIELD, key);
+        });
+  }
+
+  /**
+   * Given 4 {@link Set}s of ids of products, categories, productTypes and custom objects, this
+   * method first checks if there is a key mapping for each id in the {@code idToKey} cache. If
+   * there exists a mapping for all the ids, the method returns a future containing the existing
+   * {@code idToKey} cache as it is. If there is at least one missing mapping, it attempts to make a
+   * GraphQL request (note: rest request for custom objects) to CTP to fetch all ids and keys of
+   * every missing product, category, productType or custom object Id in a combined request. For
+   * each fetched key/id pair, the method will insert it into the {@code idToKey} cache and then
+   * return the cache in a {@link CompletableFuture} after the request is successful.
+   *
+   * @param productIds the product ids to find a key mapping for.
+   * @param categoryIds the category ids to find a key mapping for.
+   * @param productTypeIds the productType ids to find a key mapping for.
+   * @param customObjectIds the custom object ids to find a key mapping for.
+   * @return a map of id to key representing products, categories, productTypes and customObjects in
+   *     the CTP project defined by the injected {@code ctpClient}.
+   */
+  @Nonnull
+  CompletableFuture<Map<String, String>> getIdToKeys(
+      @Nonnull final Set<String> productIds,
+      @Nonnull final Set<String> categoryIds,
+      @Nonnull final Set<String> productTypeIds,
+      @Nonnull final Set<String> customObjectIds) {
+
+    final Set<String> nonCachedProductIds = getNonCachedReferenceIds(productIds);
+    final Set<String> nonCachedCategoryIds = getNonCachedReferenceIds(categoryIds);
+    final Set<String> nonCachedProductTypeIds = getNonCachedReferenceIds(productTypeIds);
+    final Set<String> nonCachedCustomObjectIds = getNonCachedReferenceIds(customObjectIds);
+
+    if (nonCachedProductIds.isEmpty()
+        && nonCachedCategoryIds.isEmpty()
+        && nonCachedProductTypeIds.isEmpty()) {
+      return fetchCustomObjectKeys(customObjectIds);
```

Comment:
Looks like, it's fetching the keys that is already in the cache: 

```suggestion
      return fetchCustomObjectKeys(nonCachedCustomObjectIds);
```